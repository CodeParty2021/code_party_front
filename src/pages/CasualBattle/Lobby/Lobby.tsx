import React from "react";
import { Link } from "react-router-dom";

type Props = {};

export const CasualBattleLobby: React.FC<Props> = () => {
  return (
    <div>
      <div>ルーム待機画面</div>
      <Link to="/casual-battle/waiting-room">ルームを建てる</Link>
      <Link to="/waiting-room">ルームを探す</Link>
    </div>
  );
};
