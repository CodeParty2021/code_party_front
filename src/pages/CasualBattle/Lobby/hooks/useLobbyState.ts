import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";
import { useEffect } from "react";
import { useNavigate } from "react-router-dom";

export type IResponse = {
  roomCreateBtnHandler: () => void;
  roomSearchBtnHandler: () => void;
};

export const useLobbyState = (): IResponse => {
  const { room, createRoom } = useRoomSync();
  const navigate = useNavigate();

  useEffect(() => {
    if (room.isEntered) {
      navigate("/casual-battle/waiting-room");
    }
  }, [room.isEntered]);

  const _roomCreateBtnHandler = () => {
    createRoom();
  };

  const _roomSearchBtnHandler = () => {
    navigate("/casual-battle/search-room");
  };

  return {
    roomCreateBtnHandler: _roomCreateBtnHandler,
    roomSearchBtnHandler: _roomSearchBtnHandler,
  };
};
