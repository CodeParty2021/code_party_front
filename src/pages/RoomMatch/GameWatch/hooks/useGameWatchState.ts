import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";

import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";
import { useFetchJson } from "hooks/ResultAPIHooks/useFetchJson";
import { useUnityGame } from "hooks/UnityGameHooks/useUnityGame";
import { UnityContext } from "react-unity-webgl";
import { useDummyLoading } from "hooks/DummyLoadingHooks/useDummyLoading";

export type IResponse = {
  unityContext: UnityContext;
  state: "Analyzing" | "Running" | "Finish" | "AnalyzingError";
  messageType: "Analyzing" | "AnalyzingError";
  exitBtnHandler: () => void;
};

export const useGameWatchState = (): IResponse => {
  const { room, updateMember } = useRoomSync();
  const { data: json, fetchJson } = useFetchJson();
  const { unityContext, startGame, unityStatus } = useUnityGame("SquarePaint");
  const [messageType, setMessageType] =
    useState<IResponse["messageType"]>("Analyzing");
  const { dummyLoadingState, startDummyLoad } = useDummyLoading(1000);
  const navigate = useNavigate();

  const state = room.info?.analyzingResult?.error
    ? "AnalyzingError"
    : !room.info?.analyzingResult?.resultId ||
      unityStatus.isLoading ||
      dummyLoadingState.isLoading
    ? "Analyzing"
    : unityStatus.isRunning
    ? "Running"
    : "Finish";

  // メッセージの操作
  useEffect(() => {
    if (state === "Analyzing") setMessageType("Analyzing");
    else if (state === "AnalyzingError") setMessageType("AnalyzingError");
  }, [state]);

  //初期処理
  useEffect(() => {
    //ユーザ状態を更新
    updateMember({
      status: "watching",
    });

    // ダミーロード開始
    startDummyLoad();
  }, []);

  // 何らかの理由でルームから出された場合はロビーに戻る
  useEffect(() => {
    if (!room.isEntered) {
      navigate("/room-match");
    }
  }, [room.isEntered]);

  //シミュレーション結果が更新されたら結果をフェッチする
  useEffect(() => {
    if (room.info?.analyzingResult?.resultId) {
      fetchJson(room.info?.analyzingResult.resultId);
    }
  }, [room.info?.analyzingResult]);

  // 観戦開始
  useEffect(() => {
    if (!unityStatus.isLoading && json) {
      startGame(json);
    }
  }, [unityStatus.isLoading, json]);

  const _exitBtnHandler = () => {
    navigate("/room-match/waiting-room");
  };

  return {
    state,
    messageType,
    unityContext,
    exitBtnHandler: _exitBtnHandler,
  };
};
