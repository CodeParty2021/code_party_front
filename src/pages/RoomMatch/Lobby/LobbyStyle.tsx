import { Sword } from "components/icons";
import styled from "styled-components";
import { GRAY_100, GRAY_40, RED_70, TURQUOISE_50 } from "styles/colors";

export const TitlePosition = styled.div`
  position: absolute;
  top: 128px;
  left: 205px;
  display: flex;
`;

export const SwordIcon = styled(Sword)`
  margin-right: 32px;
`;

export const Container = styled.div`
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translateY(-50%) translateX(-50%);
  transform: translateY(-50%) translateX(-50%);
`;

export const Card = styled.div`
  display: flex;
  width: 1511px;
  border-radius: 32px;
  margin: 51px;
  background-color: white;
`;

export const AlgoStarBanner = styled.img`
  width: 560px;
  height: 240px;
  border-radius: 32px 0 0 32px;
`;

export const AlgoGatherBanner = styled.img`
  border-radius: 32px 0 0 32px;
  width: 560px;
  height: 368px;
`;

export const ContentsArea = styled.div`
  align-items: center;
  flex-grow: 1;
  margin: auto 52px auto 52px;
  border-radius: 0 32px 32px 0;
`;

export const WithButton = styled.div`
  display: flex;
  justify-content: space-between;
`;

export const Title = styled.h2`
  color: ${GRAY_100};
  font-size: 30px;
  font-weight: 700px;
  font-feature-settings: "palt" on;
  transform: matrix(1, 0, -0.08, 1, 0, 0);
  line-height: 30px;
  display: flex;
  align-items: center;
`;

export const Form = styled.input`
  width: 663px;
  height: 76px;
  border: 2px solid ${GRAY_40};
  border-radius: 12px;
  padding: 24px;
  font-size: 24px;
  &:focus {
    border: 2px solid ${TURQUOISE_50};
  }
  &::placeholder {
    font-family: "Noto Sans JP";
    font-style: normal;
    font-weight: 700;
    font-size: 24px;
    line-height: 150%;
    color: ${GRAY_40};
  }
`;

export const Description = styled.p`
  margin: 32px 0;
  font-family: "Noto Sans JP";
  font-style: normal;
  font-weight: 400;
  font-size: 24px;
  line-height: 35px;
`;

export const ErrorMessage = styled.span`
  color: ${RED_70};
  font-size: 24px;
  font-weight: bold;
  margin-bottom: 30px;
`;
