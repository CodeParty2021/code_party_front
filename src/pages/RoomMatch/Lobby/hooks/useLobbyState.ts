import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";
import { RefObject, useCallback, useEffect, useMemo, useRef, useState } from "react";
import { useNavigate } from "react-router-dom";

export type IResponse = {
  isLoading: boolean;
  errorMessage: string;
  roomIdRef: RefObject<HTMLInputElement>;
  createRoomHandler: () => void;
  createRoomDisabled: boolean;
  enterRoomHandler: () => Promise<void>;
  enterRoomDisabled: boolean;
  backButtonHandler: () => void;
};

export const useLobbyState = (): IResponse => {
  const [isLoading, setIsLoading] = useState<boolean>(false);
  const [errorMessage, setErrorMessage] = useState<string>("");
  const [isProcessingAny, setIsProcessingAny] = useState<boolean>(false);
  const [succeedEnterRoom, setSucceedEnterRoom] = useState<boolean>(false);
  const { room, createRoom, enterRoom } = useRoomSync();
  const navigate = useNavigate();
  const roomIdRef = useRef<HTMLInputElement>(null!);

  const createEnterButtonDisabled = useMemo(() => (
    isProcessingAny || room.isEntered || succeedEnterRoom
  ), [isProcessingAny, room.isEntered]);

  useEffect(() => {
    if (room.isEntered) {
      navigate("/room-match/waiting-room");
    }
  }, [room.isEntered]);

  // ルーム
  const _createRoomHandler = useCallback(async () => {
    setIsLoading(true);
    setIsProcessingAny(true);
    await createRoom()
      .then(() => {
        setSucceedEnterRoom(true);
      })
      .catch(() => {
        navigate("/error");
      });
    setIsProcessingAny(false);
  }, []);

  const _enterRoomHandler = useCallback(async () => {
    const value = roomIdRef.current?.value;

    setIsProcessingAny(true);
    if (typeof value === "string") {
      if (value == "") {
        setErrorMessage("値を入力してください。");
      } else {
        await new Promise((resolve) => setTimeout(resolve, 500));
        await enterRoom(value).then(() => {
          setSucceedEnterRoom(true);
        }).catch((e) => {
          if (e.message == "roomId is empty") {
            setErrorMessage("値を入力してください。");
          } else if (e.message == "roomId is not found") {
            setErrorMessage("ルームIDが無効です。");
          } else if (e.message == "room is full") {
            setErrorMessage("ルームが満員です。");
          }
        });
      }
    } else {
      setErrorMessage("入力が不正です。");
    }
    setIsProcessingAny(false);
  }, []);

  // 戻るボタン
  const _backButtonHandler = useCallback(() => {
    navigate("/select-mode");
  }, []);

  return {
    isLoading,
    errorMessage,
    roomIdRef,
    createRoomHandler: _createRoomHandler,
    createRoomDisabled: createEnterButtonDisabled,
    enterRoomHandler: _enterRoomHandler,
    enterRoomDisabled: createEnterButtonDisabled,
    backButtonHandler: _backButtonHandler,
  };
};
