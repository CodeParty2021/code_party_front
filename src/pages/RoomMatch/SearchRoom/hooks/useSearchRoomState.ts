import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";

import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";

export type IResponse = {
  roomIdTextBoxValue: string;
  roomIdTextBoxChangeHandler: (value: string) => void;
  enterBtnDisabled: boolean;
  enterBtnClickHandler: () => void;
};

export const useSearchRoomState = (): IResponse => {
  const { room, enterRoom } = useRoomSync();
  const [roomId, setRoomId] = useState("");
  const [enterBtnDisabled, setEnterBtnDisabled] = useState(false);
  const navigate = useNavigate();

  useEffect(() => {
    if (room.isEntered) {
      navigate("/room-match/waiting-room");
    }
  }, [room.isEntered]);

  const _roomIdTextBoxChangeHandler = (value: string) => {
    setRoomId(value);
  };

  const _enterBtnHandler = () => {
    setEnterBtnDisabled(true);
    enterRoom(roomId).catch(() => {
      setEnterBtnDisabled(false);
    });
  };

  return {
    roomIdTextBoxValue: roomId,
    roomIdTextBoxChangeHandler: _roomIdTextBoxChangeHandler,
    enterBtnDisabled: enterBtnDisabled,
    enterBtnClickHandler: _enterBtnHandler,
  };
};
