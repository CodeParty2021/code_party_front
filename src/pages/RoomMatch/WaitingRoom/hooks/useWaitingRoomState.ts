import { useEffect, useState, useMemo } from "react";
import { useNavigate, useSearchParams } from "react-router-dom";

import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";
import { CodeType, useFetchCodes } from "hooks/CodeAPIHooks/useFetchCodes";
import { UserState, UserAction } from "services/RoomSync/RoomSync";
import { RootState } from "store";
import { useSelector } from "react-redux";
import { User } from "services/user/user";

export type IResponse = {
  user: User | null;
  roomInfo: {
    roomId: string;
    invitationLink: string;
    host: UserState;
    memberKeys: string[];
    members: { [id: string]: UserState };
    actionKeys: string[];
    actions: { [id: string]: UserAction };
  };
  isHost: boolean;
  status?: "waiting" | "watching";
  ready: boolean;

  readyBtnHandler: () => void;
  readyBtnDisabled: boolean;
  exitBtnHandler: () => void;
  startBtnDisabled: boolean;
  startBtnHandler: () => void;
  isCopyBtnClicked: boolean;
  invitationBtnHandler: () => void;
  kickUserHandler: (userId: string) => void;
  code: {
    codes: CodeType[];
    loading: boolean;
  };
  selectedCode: CodeType | null;
  onChangeSelectedCode: (code: CodeType) => void;
  onStartToEditCode: () => void;
  showCodeSelectModal: boolean;
  openCodeSelectModal: () => void;
  closeCodeSelectModal: () => void;
};

export const useWaitingRoomState = (): IResponse => {
  const {
    room,
    isHost,
    updateRoomInfo,
    updateMember,
    updateOtherMember,
    exitRoom,
  } = useRoomSync();
  const { data: codes = [], loading } = useFetchCodes();
  const [preRoomStatus, setPreRoomStatus] = useState<"waiting" | "watching">(
    "watching"
  );
  const [selectedCode, setSelectedCode] = useState<CodeType | null>(null);
  const [ready, setReady] = useState(false);
  const [readyBtnDisabled, setReadyBtnDisabled] = useState(true);
  const [startBtnDisabled, setStartBtnDisabled] = useState(true);
  const [isCopyBtnClicked, setIsCopyBtnClicked] = useState(false);
  let { user } = useSelector((state: RootState) => state.user);
  const navigate = useNavigate();
  const [searchParams] = useSearchParams();
  const dummyUser: UserState = {
    displayName: "",
    status: "disconnect",
    ready: false,
  };

  const modalOn = useMemo(
    () => (searchParams.get("modal") === "on" ? true : false),
    []
  );

  // ユーザ状態を更新
  useEffect(() => {
    updateMember({
      status: "waiting",
    });
  }, []);

  // DB上で退出処理が完了したら退出
  useEffect(() => {
    if (!room.isEntered) {
      navigate("/room-match");
    }
  }, [room.isEntered]);

  // 準備ボタンの有効化
  useEffect(() => {
    //コードが選択されたら準備完了ボタンを押せるようにする
    if (selectedCode != null) {
      setReadyBtnDisabled(false);
    }
  }, [selectedCode]);

  // ルームステータスがwaitingからwatchingになったら観戦画面に遷移
  useEffect(() => {
    if (room.info) {
      if (preRoomStatus == "waiting" && room.info.status == "watching") {
        navigate("/room-match/result");
      }
      setPreRoomStatus(room.info?.status);
    }
  }, [room.info?.status]);

  // ready状態をDBに反映
  useEffect(() => {
    updateMember({
      ready: ready,
      codeId: selectedCode ? selectedCode.id : "",
    });
  }, [ready]);

  // クエリパラメータにmodal=onを渡されたらモーダルを表示
  useEffect(() => {
    if (modalOn) {
      setShowCodeSelectModal(true);
    }
  }, [modalOn]);

  // ホストの処理
  useEffect(() => {
    if (isHost) {
      if (room.info && room.info.status == "waiting") {
        // 全員が準備完了になったら開始ボタンを有効にする
        let newDisabled = false;
        for (let key of Object.keys(room.members)) {
          if (!room.members[key].ready) {
            newDisabled = true;
            break;
          }
        }
        setStartBtnDisabled(newDisabled);
      }
      // 全員がWaitingRoomに返ってきたらroomの状態をwaitingに更新する
      let newRoomStatus: "waiting" | "watching" = "waiting";
      for (let key of Object.keys(room.members)) {
        if (room.members[key].status == "watching") {
          newRoomStatus = "watching";
        }
      }
      if (newRoomStatus == "waiting" && room.info?.status == "watching") {
        updateRoomInfo({
          status: "waiting",
          analyzingResult: null,
        });
      }
    }
  }, [room.members, isHost]);

  // kick処理 member.status = kickingの場合そのユーザをキックする
  useEffect(() => {
    if (user) {
      for (let key of Object.keys(room.members)) {
        if (room.members[key].status == "kicking" && user.id == key) {
          exitRoom();
        }
      }
    }
  }, [room.members]);

  // ---- イベントハンドラ ---- //

  const _readyBtnHandler = () => {
    setReady(!ready);
  };

  const _exitBtnHandler = () => {
    exitRoom();
  };

  const _startBtnHandler = () => {
    updateRoomInfo({
      status: "watching",
    });
  };

  const _onChangeSelectedCode = (code: CodeType) => {
    setSelectedCode(code);
  };

  const _onStartToEditCode = () => {
    if (selectedCode) {
      navigate(`/free-coding/${selectedCode.id}/selectCode`);
    }
  };

  //invitationURLのコピーボタン
  const _invitationBtnHandler = () => {
    if (room.invitationLink) {
      navigator.clipboard.writeText(room.invitationLink).then(() => {
        setIsCopyBtnClicked(true);
      });
    }
  };
  const _kickUserBtnHandler = (userId: string) => {
    updateOtherMember(userId, { status: "kicking" });
  };

  //モーダル表示非表示
  const [showCodeSelectModal, setShowCodeSelectModal] =
    useState<boolean>(false);
  const openCodeSelectModal = () => {
    setShowCodeSelectModal(true);
  };
  const closeCodeSelectModal = () => {
    setShowCodeSelectModal(false);
  };

  return {
    user: user,
    roomInfo: {
      roomId: room.id ? room.id : "",
      invitationLink: room.invitationLink ? room.invitationLink : "",
      host:
        room.info && room.info.host in room.members
          ? room.members[room.info.host]
          : dummyUser,
      memberKeys: room.sortedKeysOfMembers,
      members: room.members,
      actionKeys: room.sortedKeysOfActions,
      actions: room.actions,
    },
    isHost: isHost,
    status: room.info?.status,
    ready: ready,
    readyBtnHandler: _readyBtnHandler,
    readyBtnDisabled: readyBtnDisabled,
    exitBtnHandler: _exitBtnHandler,
    startBtnDisabled: startBtnDisabled,
    startBtnHandler: _startBtnHandler,
    isCopyBtnClicked: isCopyBtnClicked,
    invitationBtnHandler: _invitationBtnHandler,
    kickUserHandler: _kickUserBtnHandler,
    code: {
      codes: codes,
      loading: loading,
    },
    selectedCode: selectedCode,
    onChangeSelectedCode: _onChangeSelectedCode,
    onStartToEditCode: _onStartToEditCode,
    showCodeSelectModal,
    openCodeSelectModal,
    closeCodeSelectModal,
  };
};
