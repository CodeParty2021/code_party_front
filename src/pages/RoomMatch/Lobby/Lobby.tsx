import React from "react";
import { useLobbyState } from "./hooks/useLobbyState";

type Props = {};

export const RoomMatchLobby: React.FC<Props> = () => {
  const { roomCreateBtnDisabled, roomCreateBtnHandler, roomSearchBtnHandler } =
    useLobbyState();

  return (
    <div>
      <div>ルーム待機画面</div>
      <button
        id="create-btn"
        onClick={roomCreateBtnHandler}
        disabled={roomCreateBtnDisabled}
      >
        ルームを建てる
      </button>
      <button id="search-btn" onClick={roomSearchBtnHandler}>
        ルームを探す
      </button>
    </div>
  );
};
