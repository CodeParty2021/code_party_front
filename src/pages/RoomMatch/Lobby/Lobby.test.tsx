import React from "react";
import { shallow } from "enzyme";

import { RoomMatchLobby } from "./Lobby";
import { IResponse, useLobbyState } from "./hooks/useLobbyState";

jest.mock("./hooks/useLobbyState");

const useSelectorMock = useLobbyState as jest.Mock;

const state: IResponse = {
  roomCreateBtnDisabled: false,
  roomCreateBtnHandler: jest.fn(),
  roomSearchBtnHandler: jest.fn(),
};

describe("<RoomMatchLobby />", () => {
  beforeEach(() => {
    useSelectorMock.mockReturnValue({
      ...state,
    });
  });
  afterEach(() => {
    jest.resetAllMocks();
  });

  it("snapshot test", () => {
    const wrapper = shallow(<RoomMatchLobby />);

    expect(wrapper.getElements()).toMatchSnapshot();
  });

  it("room create button test", () => {
    const wrapper = shallow(<RoomMatchLobby />);

    const btn = wrapper.find("#create-btn");
    btn.simulate("click");

    expect(state.roomCreateBtnHandler).toHaveBeenCalled();
  });

  it("room search button test", () => {
    const wrapper = shallow(<RoomMatchLobby />);

    const btn = wrapper.find("#search-btn");
    btn.simulate("click");

    expect(state.roomSearchBtnHandler).toHaveBeenCalled();
  });
});
