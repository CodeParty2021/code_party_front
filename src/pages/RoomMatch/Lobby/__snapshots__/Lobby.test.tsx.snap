// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RoomMatchLobby /> snapshot test 1`] = `
Array [
  <StarBackground>
    <Header
      backButtonHandler={[Function]}
      backMessage="ゲーム選択へ戻る"
    />
    <styled.div>
      <Styled(Sword) />
      <styled.h2>
        フレンド対戦モード
      </styled.h2>
    </styled.div>
    <styled.div>
      <styled.div>
        <styled.img
          src="/img/algo_star.svg"
        />
        <styled.div>
          <styled.div>
            <styled.h2>
              新しくルームを作成する
            </styled.h2>
            <Button
              color="blue"
              onClick={[Function]}
              size="M"
              status="default"
              value="作成"
            />
          </styled.div>
        </styled.div>
      </styled.div>
      <styled.div>
        <styled.img
          src="/img/algo_gather.svg"
        />
        <styled.div>
          <styled.h2>
            ルームに参加する
          </styled.h2>
          <styled.p>
            フレンドから教えてもらったルームIDを入力して参加
          </styled.p>
          <styled.div>
            <styled.input
              placeholder="ルームIDを入力"
              type="text"
            />
            <Button
              color="blue"
              onClick={[Function]}
              size="M"
              status="default"
              value="入室"
            />
          </styled.div>
          <styled.span>
            aaa
          </styled.span>
        </styled.div>
      </styled.div>
    </styled.div>
  </StarBackground>,
]
`;
